### PR TITLE
Ensure schema change before checking OracleConstraints

### PR DESCRIPTION
**Pull request by skjnldsv**
- Fix the oracle constraints check
- Fix ldap migration parametters
- Improve error logging
  ```bash
  $ php80 occ upgrade
  Nextcloud or one of the apps require upgrade - only a limited number of commands are available
  You may use your browser or the occ upgrade command to do the upgrade
  Setting log level to debug
  Updating database schema
  Updated database
  Updating <user_ldap> ...
  Exception: Database error when running migration 1142Date20220410143802 for app user_ldap
  Column "oc_ldap_group_mapping"."ldap_dn" is type String, but exceeding the 4.000 length limit.
  Update failed
  Maintenance mode is kept active
  Resetting log level
  ```

### CODE REVIEW

**Review by come-nc — APPROVED** [review]

**Review by blizzz — APPROVED** [review]
LGTM + unblocks an upgrade issue of the targeted scenario

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php [hunk 1]
```diff
@@ -129,7 +129,7 @@ protected function getSelectQuery(string $table): IQueryBuilder {
 		$qb = $this->dbc->getQueryBuilder();
 		$qb->select('owncloud_name', 'directory_uuid')
 			->from($table)
-			->where($qb->expr()->gt($qb->func()->octetLength('owncloud_name'), '64', IQueryBuilder::PARAM_INT));
+			->where($qb->expr()->gt($qb->func()->octetLength('owncloud_name'), $qb->createNamedParameter('64'), IQueryBuilder::PARAM_INT));
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:132]: 🙈

File: lib/private/DB/MigrationService.php [hunk 2]
```diff
@@ -583,20 +582,30 @@ public function ensureOracleConstraints(Schema $sourceSchema, Schema $targetSche
...
-				if ((!$sourceTable instanceof Table || !$sourceTable->hasColumn($thing->getName())) && $thing->getNotnull() && $thing->getDefault() === ''
-					&& $sourceTable instanceof Table && !$sourceTable->hasColumn($thing->getName())) {
-					throw new \InvalidArgumentException('Column "' . $table->getName() . '"."' . $thing->getName() . '" is NotNull, but has empty string or null as default.');
-				}
+				// If the table doesn't exists OR if the column doesn't exists in the table
+				if (!$sourceTable instanceof Table || !$sourceTable->hasColumn($thing->getName())) {
+					if (\strlen($thing->getName()) > 30) {
+						throw new \InvalidArgumentException('Column name "' . $table->getName() . '"."' . $thing->getName() . '" is too long.');
+					}
+
```
Comment (nickvergessen) [inline lib/private/DB/MigrationService.php:590]:
```suggestion

```

File: lib/private/DB/MigrationService.php [hunk 3]
```diff
@@ -583,20 +582,30 @@ public function ensureOracleConstraints(Schema $sourceSchema, Schema $targetSche
...
+				if (!$sourceTable instanceof Table || !$sourceTable->hasColumn($thing->getName())) {
+					if (\strlen($thing->getName()) > 30) {
+						throw new \InvalidArgumentException('Column name "' . $table->getName() . '"."' . $thing->getName() . '" is too long.');
+					}
+	
+					if ($thing->getNotnull() && $thing->getDefault() === ''
+						&& $sourceTable instanceof Table && !$sourceTable->hasColumn($thing->getName())) {
+						throw new \InvalidArgumentException('Column "' . $table->getName() . '"."' . $thing->getName() . '" is NotNull, but has empty string or null as default.');
+					}
+
```
Comment (nickvergessen) [inline lib/private/DB/MigrationService.php:595]:
```suggestion

```

File: lib/private/DB/MigrationService.php [hunk 4]
```diff
@@ -583,20 +582,30 @@ public function ensureOracleConstraints(Schema $sourceSchema, Schema $targetSche
...
 
-				if ((!$sourceTable instanceof Table || !$sourceTable->hasColumn($thing->getName())) && $thing->getNotnull() && $thing->getType()->getName() === Types::BOOLEAN) {
-					throw new \InvalidArgumentException('Column "' . $table->getName() . '"."' . $thing->getName() . '" is type Bool and also NotNull, so it can not store "false".');
+					$sourceColumn = null;
+				} else {
+					$sourceColumn = $sourceTable->getColumn($thing->getName());
 				}
-
-				if ($thing->getLength() > 4000 && $thing->getType()->getName() === Types::STRING) {
+
```
Comment (nickvergessen) [inline lib/private/DB/MigrationService.php:604]:
```suggestion

```

File: lib/private/DB/MigrationService.php [hunk 5]
```diff
@@ -583,20 +582,30 @@ public function ensureOracleConstraints(Schema $sourceSchema, Schema $targetSche
...
 			foreach ($table->getColumns() as $thing) {
-				if ((!$sourceTable instanceof Table || !$sourceTable->hasColumn($thing->getName())) && \strlen($thing->getName()) > 30) {
-					throw new \InvalidArgumentException('Column name "' . $table->getName() . '"."' . $thing->getName() . '" is too long.');
-				}
-
-				if ((!$sourceTable instanceof Table || !$sourceTable->hasColumn($thing->getName())) && $thing->getNotnull() && $thing->getDefault() === ''
-					&& $sourceTable instanceof Table && !$sourceTable->hasColumn($thing->getName())) {
-					throw new \InvalidArgumentException('Column "' . $table->getName() . '"."' . $thing->getName() . '" is NotNull, but has empty string or null as default.');
-				}
+				// If the table doesn't exists OR if the column doesn't exists in the table
```
Comment (nickvergessen) [inline lib/private/DB/MigrationService.php:585]:
```suggestion
				// If the table doesn't exist OR if the column doesn't exist in the table
```

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**